src/sctools/count.py

```diff
@@ -0,0 +1,163 @@
+"""
+
+from typing import List, Dict, Tuple
+
+import numpy as np
+import scipy.sparse as sp
+import pysam
+
+from sctools import gtf
+
+
+def bam_to_count(
+        bam_file: str,
+        annotation_file: str,
+        cell_barcode_tag: str='CB',
+        molecule_barcode_tag: str='UB',
+        gene_id_tag: str='GE',
+        open_mode: str='rb',
+        ) -> sp.csr_matrix:
+    """Generate a count matrix from a sorted, tagged bam file
+
+    Input bam file must be sorted by cell, molecule, and gene (where the gene tag varies fastest).
+    This module returns reads that correspond to both spliced and unspliced reads.
+
+    Parameters
+    ----------
+    bam_file : str
+        input bam file marked by cell barcode, molecule barcode, and gene ID tags sorted in that
+        order
+    cell_barcode_tag : str, optional
+        Tag that specifies the cell barcode for each read. Reads without this tag will be ignored
+        (default = 'CB')
+    molecule_barcode_tag : str, optional
+        Tag that specifies the molecule barcode for each read. Reads without this tag will be
+        ignored (default = 'UB')
+    gene_id_tag
+        Tag that specifies the gene for each read. Reads without this tag will be ignored
+        (default = 'GE')
+    annotation_file : str
+        gtf annotation file that was used to create gene ID tags. Used to map genes to indices
+    open_mode : {'r', 'rb'}, optional
+        indicates that the passed file is a bam file ('rb') or sam file ('r') (default = 'rb').
+
+    Returns
+    -------
+    count_matrix : sp.csr_matrix
+        cells x genes sparse count matrix in compressed sparse row format (cells are compressed)
+
+    Notes
+    -----
+    Any matrices produced by this function that share the same annotation file can be concatenated
+    using the scipy sparse vstack function, for example:
+
+    >>> import scipy.sparse as sp
+    >>> A = sp.coo_matrix([[1, 2], [3, 4]]).tocsr()
+    >>> B = sp.coo_matrix([[5, 6]]).tocsr()
+    >>> sp.vstack([A, B]).toarray()
+    array([[1, 2],
+           [3, 4],
+           [5, 6]])
+
+    See Also
+    --------
+    samtools sort (-t parameter):
+        C library that can sort files as required.
+        http://www.htslib.org/doc/samtools.html#COMMANDS_AND_OPTIONS
+    TagSortBam.CellSortBam:
+        WDL task that accomplishes the sorting necessary for this module.
+        https://github.com/HumanCellAtlas/skylab/blob/master/library/tasks/TagSortBam.wdl
+
+    """
+
+    # create input arrays
+    data: List[int] = []
+    cell_indices: List[int] = []
+    gene_indices: List[int] = []
+
+    gene_id_to_index: Dict[str, int] = {}
+    gtf_reader = gtf.Reader(annotation_file)
+
+    # map the gene from reach record to an index in the sparse matrix
+    for gene_index, record in enumerate(gtf_reader.filter(retain_types=['gene'])):
+        try:
+            gene_id = record.get_attribute('gene_id')
+        except KeyError:
```

Comment:
I have a feeling that you could only get a None instead of a KeyError from using this method: https://github.com/HumanCellAtlas/sctools/blob/53b22c935bd537ce901db541899ca7d5f249c01f/src/sctools/gtf.py#L146, since it is `dic.get('key')` instead of `dic['key']`. I could be wrong though.

Reply:
Nice 🐛